docs/development/testing.md

```diff
@@ -0,0 +1,166 @@
+
+* Creating sample data for tests can be a high effort. Consider writing test utilities to generate sample data instead. [example-test-object-builder](https://github.com/gardener/etcd-druid/blob/4e9971aba3c3880a4cb6583d05843eabb8ca1409/test/utils/etcd.go#L61).
+
+* If tests require any arbitrary sample data then ensure that you create a `testdata` directory within the package and keep the sample data as files in it. From https://pkg.go.dev/cmd/go/internal/test
+
+  > The go tool will ignore a directory named "testdata", making it available to hold ancillary data needed by the tests.
+
+* Avoid defining shared variable/state across tests. This can lead to race conditions causing non-deterministic state. Additionally it limits the capability to run tests concurrently via `t.Parallel()`.
+
+* Do not assume or try and establish an order amongst different tests. This leads to brittle tests as the codebase evolves.
+
+* If you need to have logs produced by test runs (especially helpful in failing tests), then consider using [t.Log](https://pkg.go.dev/testing#T.Log) or [t.Logf](https://pkg.go.dev/testing#T.Logf).
+
+## Unit Tests
+
+* If you need a kubernetes `client.Client`, prefer using [fake client](https://github.com/gardener/etcd-druid/blob/master/test/utils/client.go#L67) instead of mocking the client. You can inject errors when building the client which enables you test error handling code paths.
+    * Mocks decrease maintainability because they expect the tested component to follow a certain way to reach the desired goal (e.g., call specific functions with particular arguments).
+* All unit tests should be run quickly. Do not use [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest) and do not set up a [Kind](https://kind.sigs.k8s.io/) cluster in unit tests.
+* If you have common setup for variations of a function, consider using [table-driven](https://go.dev/wiki/TableDrivenTests) tests. See [this](https://github.com/gardener/etcd-druid/blob/4e9971aba3c3880a4cb6583d05843eabb8ca1409/internal/component/rolebinding/rolebinding_test.go#L27) as an example.
+* An individual test should only test one and only one thing. Do not try and test multiple variants in a single test. Either use [table-driven](https://go.dev/wiki/TableDrivenTests) tests or write individual tests for each variation.
+* If a function/component has multiple steps, its probably better to split/refactor it into multiple functions/components that can be unit tested individually.
+* If there are a lot of edge cases, extract dedicated functions that cover them and use unit tests to test them.
+
+### Running Unit Tests
+
+> **NOTE:** For unit tests we are currently transitioning away from [ginkgo](https://github.com/onsi/ginkgo) to using golang native tests. The `make test-unit` target runs both ginkgo and golang native tests. Once the transition is complete this target will be simplified.
+
+Run all unit tests
+
+```bash
+> make test-unit
```

Comment:
This is not possible right now. It is because we are using a tool to beautify the output which otherwise is harder to read. In the current output format the coverage is very minimally included as % against a package. So even if i introduce 2 variants one with coverage and one without the difference in verbosity will not be much, at least not enough to trouble you.

I will be replacing this tool with something that i will develop separately. I do not like any tool that is currently available in the community. Ginkgo uses a formatter which they have internally developed. We currently cannot use this as it is not an independent component.